{{cookiecutter.project_slug}}/config/asgi.py

```diff
@@ -0,0 +1,44 @@
+"""
+ASGI config for {{ cookiecutter.project_name }} project.
+
+It exposes the ASGI callable as a module-level variable named ``application``.
+
+For more information on this file, see
+https://docs.djangoproject.com/en/dev/howto/deployment/asgi/
+
+"""
+import os
+import sys
+from pathlib import Path
+
+from django.core.asgi import get_asgi_application
+
+# This allows easy placement of apps within the interior
+# {{ cookiecutter.project_slug }} directory.
+app_path = Path(__file__).parents[1].resolve()
+sys.path.append(str(app_path / "{{ cookiecutter.project_slug }}"))
+# We defer to a DJANGO_SETTINGS_MODULE already in the environment. This breaks
+# if running multiple sites in the same mod_wsgi process. To fix this, use
+# mod_wsgi daemon mode with each site in its own daemon process, or use
+# os.environ["DJANGO_SETTINGS_MODULE"] = "config.settings.production"
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "config.settings.production")
+
+# This application object is used by any ASGI server configured to use this
+# file. This includes Django's development server, if the ASGI_APPLICATION
+# setting points here.
+django_application = get_asgi_application()
+# Apply ASGI middleware here.
+# from helloworld.asgi import HelloWorldApplication
+# application = HelloWorldApplication(application)
+
+# Import websocket application here, so apps from django_application are loaded first
+from .websocket import websocket_application  # noqa isort:skip
```

Comment:
I don't understand why we cannot import it at the top, the `websocket_application` doesn't require the application to be loaded, does it? Or is it because we can expect that it _will_ as soon as some meaningful code is in there?

Reply:
I believe from past experience when importing code from the app, I would always run into errors like apps weren't loaded. When I created a chat app with Redis Pub/Sub, especially, I had to access celery tasks with `asgiref.sync.sync_to_async` to run some db commands (also possible with database_sync_to_async from Django channels) and also access some code from other places. Put it simply, it's best if the import is left there.